kernel/src/hil/time.rs

```diff
@@ -77,24 +86,92 @@ pub trait Alarm: Time {
+
+    /// Returns whether this alarm is currently active (will eventually trigger
+    /// a callback if there is a client).
+    fn is_enabled(&self) -> bool;
+
+    /// Enables the alarm using the previously set `tics` for the alarm.
+    ///
+    /// Most implementations should use the default implementation which calls `set_alarm` with the
+    /// value returned by `get_alarm` unless there is a more efficient way to achieve the same
+    /// semantics.
+    fn enable(&self) {
+        self.set_alarm(self.get_alarm())
+    }
+
+    /// Disables the alarm.
+    ///
+    /// The implementation will _always_ disable the alarm and prevent events related to previously
+    /// set alarms from being delivered to the client.
+    fn disable(&self);
 }
 
-/// A client of an implementor of the [`Alarm`](trait.Alarm.html) trait.
-pub trait Client {
+/// A client of an implementer of the [`Alarm`](trait.Alarm.html) trait.
+pub trait AlarmClient {
     /// Callback signaled when the alarm's clock reaches the value set in
     /// [`Alarm#set_alarm`](trait.Alarm.html#tymethod.set_alarm).
     fn fired(&self);
 }
 
 /// The `Timer` trait models a timer that can notify when a particular interval
 /// has elapsed.
-pub trait Timer: Time {
+pub trait Timer<'a, W = u32>: Time<W> {
+    /// Set the client for interrupt events.
+    fn set_client(&'a self, client: &'a TimerClient);
+
     /// Sets a one-shot timer to fire in `interval` clock-tics.
-    fn oneshot(&self, interval: u32);
+    ///
+    /// Calling this method will override any existing oneshot or repeating timer.
+    fn oneshot(&self, interval: W);
+
     /// Sets repeating timer to fire every `interval` clock-tics.
-    fn repeat(&self, interval: u32);
+    ///
+    /// Calling this method will override any existing oneshot or repeating timer.
+    fn repeat(&self, interval: W);
+
+    /// Returns the interval for a repeating timer.
+    ///
+    /// Returns `None` if the timer is disabled or in oneshot mode and `Some(interval)` if it is
+    /// repeating.
+    fn interval(&self) -> Option<W>;
+
+    /// Returns whether this is a oneshot (rather than repeating) timer.
+    fn is_oneshot(&self) -> bool {
+        self.interval().is_none()
+    }
+
+    /// Returns whether this is a repeating (rather than oneshot) timer.
+    fn is_repeating(&self) -> bool {
+        self.interval().is_some()
+    }
+
+    /// Returns the remaining time in clock tics for a oneshot or repeating timer.
+    ///
+    /// Returns `None` if the timer is disabled.
+    fn time_remaining(&self) -> Option<W>;
+
+    /// Returns whether this timer is currently active (has time remaining).
+    fn is_enabled(&self) -> bool {
+        self.time_remaining().is_some()
+    }
+
+    /// Cancels an outstanding timer.
+    ///
+    /// The implementation will _always_ cancel the timer.
+    /// delivered.
+    fn cancel(&self);
+}
+
+/// A client of an implementer of the [`Timer`](trait.Timer.html) trait.
```

Comment:
Why have separate Alarm and Timer clients?

Reply:
The argument was that (a) Alarm and Timer are different traits so they should have different client types, and (b) this is the way for a client that _does_ happen to be use/need both functionalities having separate client traits makes it easy to distinguish which event it just got.